### PR TITLE
[WIP] Unification of timeout(..) and after(..) API aka within()   

### PR DESCRIPTION
**Pull request by ChristianSchwarz**
----
**Please use this issue to review only the implementation. For API related discussions like why within() and not xy() please use ticket [#472 Unify after() and timeout() ](../issues/472) !**

----

__**Fixes**: [#472 Unify after() and timeout() ](../issues/472)__
Fixes: [#417 Add new WithinRange VerificationMode](../issues/417)
Fixes: [#345 Verification using After and ArgumentCaptor returns unexpected size of captured values list](../issues/345)
Fixes: [#1037 Duplicate arguments captured with multithreaded method call](../issues/1037)
Relates to: [#91 Document Mockito.after() method ](../issues/91)

### What
* This PR unifies the API of `timeout()` and `after()` into the new API `within()` that is accessible throu `Mockito.within(..)`. 
* This PR also includes an enhanced `atLeast()` API that allow to combine it with `atMost()` (see #417). 
Example:  `verify(mock, within(100,MILLISECONDS).atLeast(2).andAtMost(4)).simpleMethod();`

### How the behaviour differs from after/timeout
A within-verification always succeed or fail fast where applicable. Not all verification can fail or succeed fast. Eg. times(n) can only fail fast if more calls are detected but not succeed fast cause it needs to wait the full time if less calls than required where detected.

* Verification modes that check for a lower  bound will succeed fast if enough calls were detected.
Example:
```java
mock.methodCall();
mock.methodCall();
mock.methodCall();
verify(mock, within(100,MILLISECONDS).atLeast(2)).methodCall()
//^ succeed immediatly, doesn't wait the full 100ms
```
* Verification modes that check for a upper bound will fail fast if more calls were detected.
Example:
```java
mock.methodCall();
mock.methodCall();
mock.methodCall();
verify(mock, within(100,MILLISECONDS).times(2))).methodCall()
//^ fails immediatly, doesn't wait the full 100ms
```

### Compability
The within implementation is fully compatible cause it doesn't change the timeout or after API nor internal classes. 

### Implementation
Unlike _after_ and _timeout_ the _within_ implementation doesn't poll for invocations in a loop but uses an queue based approach. When the verification starts all invocations that where made until now where placed in that queue. New invocations are added to the queue using an invocation listener. The queue will be polled and its invocations are passed to the verification strategy (e.g. never, atMost ,..) it decides if the verification condition is satified or the verification failed. This allows to fail or succeed fast.

### Current state
The following verifications are working now:
 * verify(atLeast(n))
 * verify(atLeast(n).atMost(m)) 
 * verfiy(atMost(n))
 * verify(only())
 * verify(times(n))
 * verify(never())
 * verify(within(duration,timeunit).atLeast(n))
 * verify(within(duration,timeunit).atLeast(n).atMost(m)) 
 * verfiy(within(duration,timeunit).atMost(n))
 * verify(within(duration,timeunit).only())
 * verify(within(duration,timeunit).times(n))
 * verify(within(duration,timeunit).never())

Verifications *InOrder are work in progress*, thats why the tests fail.

### CODE REVIEW

**Comment by ChristianSchwarz** [comment]
@core-team
Does it make sense to continue work on this PR or is it unlikely that this new API find its way into mockito 2?

**Comment by TimvdLippe** [comment]
Personally I have not yet had the team to review the current API, the problems and how we can solve them. Unsure at this moment whether we will merge it or not, not saying it will definitely not.

**Comment by ChristianSchwarz** [comment]
@TimvdLippe Just for reference, the API-discussion can be found here -> [#472 Unify after() and timeout() ](../issues/472)

**Comment by szczepiq** [comment]
>Does it make sense to continue work on this PR or is it unlikely that this new API find its way into mockito 2?

Given that changes are compatible (deprecations are ok but be gentle for our users :) I see no problem merging it in 2.x. I didn't have time to review the code changes yet.

The description of this PR is very informative, thank you. To make the review faster and improve the documentation even better could you make it crystal clear in what way the APIs are changed:
- update the description and add *before* and *after* state of various verification modes (or perhaps annotate which modes are new)
- any deprecations?
- the link to discussion in #472 is not super useful because the target thread is long and hard to follow. Can we update the description of #472 or summarize it here?

From the description I am super excited about this feature. Thank you very much for picking this up and THANK you for your patience waiting for review. We will get to it!

**Comment by ChristianSchwarz** [comment]
@szczepiq 
* I updated the description of this PR to describe how it differs from timeout/after. 
* no deprecations are planned for this PR, but i can deprecate after/timeout if requested
* compability: the current API of timeout/after is untouched also the internal implementation, all my additions life in package  org.mockito.internal.verification.within
* if something is unclear or questionable let me know

**Comment by szczepiq** [comment]
@ChristianSchwarz, thank you very much for updating the PR description. It is excellent!!!

+1 for going ahead

**Comment by szczepiq** [comment]
@ChristianSchwarz, is this ready for review? I'm excited about this change!

**Comment by ChristianSchwarz** [comment]
@szczepiq Yes and No!
* Yes -> The currently implemented features are ready for a review there are is no more stuff i see that need to be refactored, but they lack some JavaDocs. 
* No -> The InOrder feature is still missing, i need figure out how the current "greedy"/"chunk" strategy works.

Anyway, I would like to get a review for the code of this PR! It is quite a lot so I we need more review iterations.

I'am quite busy with my private stuff, so I think the  missing InOrder stuff must wait until the end of the summer.

**Comment by ChristianSchwarz** [comment]
Hi @TimvdLippe and @szczepiq I know you guys are buzy! Any kind of feedback/review on this PR is welcome. If you need some more infomations about implementation details let me know.

**Comment by TimvdLippe** [comment]
PR design looks great, but sadly I don't have the time to review 1400+ lines of code :cry: Hopefully someone else has time to review

**Comment by ChristianSchwarz** [comment]
Yes this is quite a lot of code, luckily +50% are JUnit-Tests mostly adapted from existing after/timeout tests. 

Here are 3 blocks that can reviewed individually, I hope this makes it a bit easier for you guys' to review the PR step by step (click on the link to see the diff):

 * Core-API's:  [Mockito](https://github.com/mockito/mockito/pull/936/files#diff-ddbc0d47e06a589778a6efe24c5d71d3),[Within ](https://github.com/mockito/mockito/pull/936/files#diff-c89535ae760e82daa1f4a94a1ef4cd1b), [VerificationModeAtLeast](https://github.com/mockito/mockito/pull/936/files#diff-faa42405ec50bab8908b5518ccd78724)
 * Core-Implementation: [VerificationStrategy](https://github.com/mockito/mockito/pull/936/files#diff-bad4c68757beab1e41c5f96836682ab1), [WithinVerification](https://github.com/mockito/mockito/pull/936/files#diff-2499b5f370c738e06a8259c1482018a2)
 * Verification Mode implementations:   [AtLeast](https://github.com/mockito/mockito/pull/936/files#diff-6dd9884084e1009b9b40dc178ea98cdc), [AtLeastAndAtMost](https://github.com/mockito/mockito/pull/936/files#diff-e8ca80b89070617c91438b8e38c360e8), [AtMost](https://github.com/mockito/mockito/pull/936/files#diff-e4d842bbed2fe68c5630a30cae3e54f7), [Only](https://github.com/mockito/mockito/pull/936/files#diff-228a0197d9c38debfc91cde6cbdda302), [Times](https://github.com/mockito/mockito/pull/936/files#diff-fcc80297d76730cee1d04dcae36a4ab0)

**Comment by ChristianSchwarz** [comment]
Ping.....

**Comment by ChristianSchwarz** [comment]
Pong...

**Comment by TimvdLippe** [comment]
I haven't reviewed this PR yet and I have been hesitating to do so. My main issue is that this introduces new public API on our pretty large public API already. In our API there are already multiple ways to achieve the same result. This PR introduces yet another instance. I do agree that we should clarify and fix any inconsistencies, but I would rather do a large analysis of our API first and then determine how to proceed. (Preferably in Mockito 3, but we haven't defined any scope for 3 just yet).

As such, I am inclined to let this PR rest for the time being, until we have resolved the overall discussion of our API and a thorough analysis of what our API contains and should contain.

**Comment by ChristianSchwarz** [comment]
Okay, thanks for taking time to give me some feedback on this PR!

I am afraid I am lost now in an infinite loop: 
> Preferably in Mockito 3, but we haven't defined any scope for 3 just yet

> https://github.com/mockito/mockito/issues/472#issuecomment-231420713
> image


----
I have spend hours of my free time to this PR, based on the feedback I requested (see #472). I wanted to avoid a situation like this. Now this PR is in question and god knows if it ever will be  accepted. 

In https://github.com/mockito/mockito/pull/936#issuecomment-283040003) @szczepiq give me a "go ahead" so I moved ahead. I can fully understand that other thigs have a higher prio now and there is no time for a review. But please give your contributors first "go ahead" and later a "maybe".

**Comment by TimvdLippe** [comment]
@ChristianSchwarz I agree this is a good change, but I rather would like to take a closer look at the full API before deciding how to proceed. The PR is in a good shape, it's just that I would like to have a better overview of what changes we are going to make. Similarly to this API, there are some APIs that have to be revised. As such, we can take everything into consideration and then take some action. For now, I would rather wait with this PR till we have resolved that discussion. Probably then we can proceed and likely merge. As we have several core developers unavailable for the time being, work has stalled for a bit. That's why we took a lot longer to review and respond. Apologies for that.

**Comment by ChristianSchwarz** [comment]
@TimvdLippe Any news?

**Comment by TimvdLippe** [comment]
@ChristianSchwarz Not as of yet. We are very low on time resources atm, so not sure if we can provide any feedback any time soon :cry:

**Comment by mockitoguy** [comment]
Since I gave you a go ahead it's on me to review it ;) I'll commit to review by 6/30.

Thank you for patience! Big changes are hard to review thoroughly. Next time, let's find an incremental way to land a big change.

**Comment by ChristianSchwarz** [comment]
@mockitoguy Thank you for looking into it! To make the review a little bit easier you may want to look at this comment for faster navigation throu the imlementation:https://github.com/mockito/mockito/pull/936#issuecomment-306466341 .

**Comment by mockitoguy** [comment]
@ChristianSchwarz, will do, thank you for your suggestions on how to review this change!

I've noticed some conflicts with our main branch. For some reason, GitHub shows clean merge. When I tried "git rebase release/2.x" from command line, I got conflicts. Can you rebase?

We'll get this change in soon!

**Comment by mockitoguy** [comment]
@ChristianSchwarz, can you summarize the state of the PR?
 - I get compilation error when I try to build your branch from command line
 - the description and comments indicate some unfinished work with "inOrder"

Thanks!

**Comment by ChristianSchwarz** [comment]
@mockitoguy 

I am going to fix the compilation errors. The `InOrder `Implementation is not ready yet, all related tests are ignored for now.

**Comment by mockitoguy** [comment]
Thank you for fixing the problems! I have a couple of questions:

- does the new 'within' API intend to deprecate the existing 'after' and 'timeout' methods?
- the description indicates that #417 is addressed, how? Mockito.atLeast().atMost() is not implemented in this PR.
- what is 'untilNow' method? It is a new public method but it is not included in PR description

> The InOrderImplementation is not ready yet, all related tests are ignored for now.

What does it mean for the end user? The current after/timeout are not available for inOrder anyway, right?

Other:

- links in PR description are broken
- there is no javadoc on many public methods and types. Don't write it yet, we need to first nail down the API we want. I just want to mention that without comprehensive Javadoc documentation / examples it is not in mergeable state.

General feedback:

- The way we evolve the API is very gradual and cautious. This means that we will have more discussions about it, including reviving old threads and reiterating past arguments. Thank you for patience!
- It seems that this is a lot of changes for fixing 2 problems reported by the users: a) lack of atLeast().atMost() - #417 - minor problem, has clean workaround , b) bugs in 'after' implementation (#345). I'm in favor of approaching it incrementally - fixing the bugs, then API rework.
- Please be constructive when working with us. Let's focus on design and code, not on complaining on slow turnaround. Fixing this problem is not high priority compared to other Mockito work (examples: Mockito 3, strictness, mockio-scala, mockito-android). Thank you for understanding!

This is a lot of hard work. Thank you for working on this with us!

**Comment by ChristianSchwarz** [comment]
> does the new 'within' API intend to deprecate the existing 'after' and 'timeout' methods?

Yes, all uses cases of `after` and `timeout `should be covered by `within`.

**Comment by ChristianSchwarz** [comment]
> the description indicates that #417 is addressed, how? Mockito.atLeast().atMost() is not implemented in this PR.

The API look likes this: `within(..).atLeast(min).andAtMost(max)` , samples can be found here: [AtLeastAndAtMostTest](https://github.com/mockito/mockito/pull/936/files#diff-aea5f220c9c98f395d1ee1533ea9e6d3).
The API implementation can be found in [Within.atLeast(..)](https://github.com/mockito/mockito/pull/936/files#diff-c89535ae760e82daa1f4a94a1ef4cd1bR56) , The underlining verification logic its composed in [AtLeastAndAtMost](https://github.com/mockito/mockito/pull/936/files#diff-e8ca80b89070617c91438b8e38c360e8). It simply combines both `AtLeast` + `AtMost`.

**Comment by ChristianSchwarz** [comment]
> what is 'untilNow' method? It is a new public method but it is not included in PR description

Thanks for the pointer I am going to remove `untilNow`! `untilNow` is equal to `within(0,SECONDS)`. It demonstrates that the `within` implementation can be used to replace the current verification mode implementations too (reduces the code complexity). 

E.g.  Mockito.times(int) can be implmented reusing the within implementation
```java
public static VerificationMode times(int wantedNumberOfInvocations) {
   return Within.untilNow().times(wantedNumberOfInvocations); 
}
```

Sorry for the confusion! In case this PR lands, the replacement of the existing implementation can be discussed again.

**Comment by ChristianSchwarz** [comment]
> >  The InOrderImplementation is not ready yet, all related tests are ignored for now.
> 
> What does it mean for the end user? The current after/timeout are not available for inOrder anyway, right?

I can remove it from the API. This simplifies the PR, the InOrder Feature can be added later if required.

**Comment by ChristianSchwarz** [comment]
>     links in PR description are broken
I fixed them last week, they work for me.

>     there is no javadoc on many public methods and types. Don't write it yet, we need to first nail down the API we want. I just want to mention that without comprehensive Javadoc documentation / examples it is not in mergeable state.

I was focused on a clean implementation and user friendly API, there was alot of back and forth. Once the general implementation and API is accepted, I am going to write the docs if that is okay for you.

**Comment by ChristianSchwarz** [comment]
> General feedback:
> 
> It seems that this is a lot of changes for fixing 2 problems reported by the users[..] I'm in favor of approaching it incrementally - fixing the bugs, then API rework.

I see you point! This PR was not intended to fix the known after/timeout bugs. `Timeout`/`after` still have them, but they don't occure in the `within` implementation.


> Please be constructive when working with us. Let's focus on design and code, not on complaining on slow turnaround. Fixing this problem is not high priority compared to other Mockito work (examples: Mockito 3, strictness, mockio-scala, mockito-android). 

I understand, thanks!

**Comment by mockitoguy** [comment]
> Yes, all uses cases of after and timeoutshould be covered by within.

@ChristianSchwarz, how so? Currently 'after' and 'timeout' serve 2 use cases that are similar but different: 'after' awaits full duration, 'timeout' returns immediately on success. Do you think that we don't need to support those 2 use cases any more?

> The API look likes this: within(..).atLeast(min).andAtMost(max)

This means that I cannot use 'atLeast' + 'atMost' without 'within'? If that's the case, are really addressing #417?

Thanks for patience, I'm working on it when I can.

**Comment by mockitoguy** [comment]
While working on this I found the existing documentation and test coverage inadequate, I submitted #1425. I will also submit a PR for fixing the tests.

**Comment by ChristianSchwarz** [comment]
@mockitoguy 

> Currently 'after' and 'timeout' serve 2 use cases that are similar but different: 'after' awaits full duration, 'timeout' returns immediately on success. Do you think that we don't need to support those 2 use cases any more?

Right, we don't need both. `Within` can fail or succeed fast, depending on the verification mode. You get the best of `timeout` and `after` in one API here. If one want to verify after full duration (without failing or succeding fast) then the JRE standard utils should be used. This was discussed in #417.

E.g.
```java
 sleep(50,MILLIS);
 verify(mock).call(123);
```

Or with Google-Guava [Uninterruptibles](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/Uninterruptibles.html#sleepUninterruptibly(long,%20java.util.concurrent.TimeUnit)), if you don't like `InterruptedException's`
```java
 sleepUninterruptibly(50,MILLIS);
 verify(mock).call(123);
```

----

> This means that I cannot use 'atLeast' + 'atMost' without 'within'? If that's the case, are really addressing #417?

Right this case is not included in this PR. It can be implemented in an follow up PR by calling `within(..)` without duration,like this:

Pseudo code:
```java 
VerificationModeAtLeast atLeast(int min){
  return within(0, SECONDS).atLeast(min)
}
```

**Comment by mockitoguy** [comment]
I made a first deep dive of the implementation and I love it! It is a great idea to avoid polling and use invocation listeners!

Some questions/feedback (for discussion, not necessarily changing code just yet):

1. Do we want to deprecate 'after()' without a replacement
2. Can we make the existing timeout() use within() behind the hood?
3. I am worried about thread safety of shared list of invocation listeners
4. I'm considering following API tweak (thoughts?)

```java
//instead:
.atLeast(10).andAtMost(20)
//making both work:
.atLeast(10).atMost(20)
.atMost(10).atLeast(20)
```

5. Thoughts about within(100) method that uses millis by default?

I really like those changes, we should have worked on them earlier. Thank you!

**Comment by ChristianSchwarz** [comment]
>  1.  Do we want to deprecate 'after()' without a replacement

I would do it, otherwise the problem of 2 almost identic API's for nearly the same uses case will remain. Otherwise mockito users will still wonder which API to use for what. 
The deprecation doc should reference to `within` and show workaround example. From my perspective as developer it should be enough. 

---

> 2.  Can we make the existing timeout() use within() behind the hood?

Good question! To be honest after more than a year I forget the details of timeout/after and my own within implementation in parts. To make a definite decission we need something like a feature matrix indicating when timeout/after succeed / fail fast and when not. If it is possible we could possibly drop the VerificationWrapper implementations. 

---

> 3.  I am worried about thread safety of shared list of invocation listeners

Thanks for that hint! I take a look.

---

> 4. I'm considering following API tweak (thoughts?)
>```java
>//instead:
>.atLeast(10).andAtMost(20)
>//making both work:
>.atLeast(10).atMost(20)
>.atMost(10).atLeast(20)
>```

If you want I can change it to the more formal `.atLeast(10).atMost(20)`.  I prepended 'atMost' with 'and' to make clear that both verifications are tied together by an _and_ operator and not by _or_. 

About `.atMost(10).atLeast(20)`: Hmm, I would like to support only one way to specifiy the low and upper bound. Otherwise tests may get mixed up with both variants depending on the preferences of devs or teams. 

> Thoughts about within(100) method that uses millis by default?
 - It is shorter.
 - The unit is not clear, cloud be seconds, nanos too. 
 -  Maybe `withinMillis(100)` as overload next to `within(int,TimeUnit)`? But then we need overloads for seconds, nanos, hours and so on too. 
 - The TimeUnit should be configurable
 - Overloads widen the API, this not required here

>I really like those changes, we should have worked on them earlier.
Glad you like it!

**Comment by mockitoguy** [comment]
This PR is still on my list. I'll get back to is shortly.

**Comment by mockitoguy** [comment]
@ChristianSchwarz, can you rebase with release/2.x?

I have good news. When I replaced the existing after() and timeout() implementation with your 'within' all tests passed :D Plus, I managed to un-ignore some existing tests that documented the desired behavior not yet implemented.

My plan is to get your within implementation into the code under the existing timeout() and after() methods. Then I'd like to revisit the public API discussion.

**Comment by mockitoguy** [comment]
PS. I pushed work to new 'ChristianSchwarz-within-sf' branch.

**Comment by ChristianSchwarz** [comment]
@mockitoguy  Wow this are really good news! I looked into your changes for after/timout and can't believe that is was that easy.

**Comment by mockitoguy** [comment]
Here's my current plan of merging this:
 - merge #1474
 - rebase this branch
 - fix the code so that the test added in #1474 works
 - remove within public API
 - merge!

It's more effort because we should clean up the inOrder implementation and add necessary testing.

**Comment by ChristianSchwarz** [comment]
@mockitoguy 

> Here's my current plan of merging this:
> * remove within public API

So `timeout()` will use the within implementation and its API (`atLeast(min).andAtMost(max)`) ?

**Comment by ChristianSchwarz** [comment]
@mockitoguy 

Should I replace the current timeout() implementation with the (new) Within implementation? If preferred I can can rename the corresponding classes and methods to Timeout/timeout.

**Comment by ChristianSchwarz** [comment]
@mockitoguy Is there any interest in this feature?

**Comment by ChristianSchwarz** [comment]
Dear core mockito members (@mockitoguy, @bric3, @TimvdLippe) I know you are all busy, nevertheless I would like to ask for a review, maybe it is possbile in 2020 or 2021. My goal it to finish it before this PR turns 5 years old.

**Comment by TimvdLippe** [comment]
This PR is large and takes a lot of time to review. Moreover, I think @mockitoguy is the best person to review this code. Overall, I am not sure if this PR is worth the effort and time required to validate the approach, for the gains it provides. I think that has been the primary factor why this PR has been left as-is.

**Comment by TimvdLippe** [comment]
Dear @ChristianSchwarz,

It's been a long time since you originally opened this PR. The effort you put in this PR undoubtedly cost you a lot of personal time and effort to get it in the state as it is. And sadly our response has been timid at best and missing at worst.

Unfortunately, the PR implements a major new API that attempts to replace existing APIs that are widely used by users. Therefore, the bar of replacing these APIs with a new one is high, if not unobtainably high. For us as maintainers, that requires a significant benefit to the new API that would otherwise be unachievable, while also the old solution being inadequate in the majority of cases.

For that last part, I think we probably come to a different conclusion than you. While there are definitely issues with `timeout` and `after`, it is my personal opinion that they are not sufficiently user breaking to warrant complete removal. Yes, they are definitely annoying and preferably our users don't run into them. But they are not part of the vast majority of cases (from anecdotal behavior) and hence (again my subjective call) do not outweigh the costs of a replacement.

This undoubtedly will be unsatisfying, especially since you put so much effort in it. The options we have now is to leave this PR open, do a thorough a review and then potentially merge it, or close it. Based on that, it is my personal perspective that I don't think PR will likely get merged for the above reasons. Therefore, rather than keeping the PR open, I think it is better to close it.

That said, I do think we should avoid a repeat of this situation, where a passionate and contributing member of the community submits a big piece of work that ends up not being included. In hindsight, it wasn't clear from the discussion in #472 that there was still a design discussion to be had. We should have probably been clearer in our communication in that the objections mentioned in the thread (e.g. new API, can we fix the old one, etc...) were objections that would reasonably block a PR implementation.

For the future, whenever there is a proposal of a major new API, I think the feedback from maintainers needs to make it clear that it will require extensive design discussion. We could look into a community structure such as RFCs to facilitate these discussions, but I am not 100% sure if extra process is going to help here.

In any case, I want to apologize again for the confusion and our hesitation to adopt this new API. The work you have put in has been significant and ideally I would have loved us to prevent significant time investment that would then not get adopted.

Sorry,
Tim
